### PR TITLE
7085 add support for "if" and "else" statements in dtrace

### PR DESCRIPTION
**Pull request by ahrens**
https://www.illumos.org/issues/7085

For those who don't like github, the code is also available here: http://cr.illumos.org/~webrev/mahrens/ifelse/

I ran the dtrace test suite (on OmniOS) and it has the same set of failures as trunk:

```
#   KIND       TEST                            DETAILS
0   cpc        tst.allcpus.ksh
1   dtraceUtil tst.ELFGenerationOut.d.ksh      dtrace: failed to link script /dev/stdin: failed to open d.out: Permission deni
2   dtraceUtil tst.ELFGenerationWithO.d.ksh    dtrace: failed to link script /dev/stdin: failed to open outputFile: Permission
3   ip         tst.ipv4localicmp.ksh
4   ip         tst.ipv4remoteicmp.ksh
5   ip         tst.ipv6remoteicmp.ksh          remote IPv6 host.
6   ip         tst.localtcpstate.ksh
7   ip         tst.remotetcpstate.ksh
8   java_api   tst.Bean.ksh                    failed to create TestBean.out
9   pid        tst.newprobes.ksh               dtrace: pid 114297 has exited
10  preprocessor err.ifdefnotendif.d
11  scripting  tst.egid.ksh                    Error in executing /var/tmp/stdin.127634.d
12  scripting  tst.euid.ksh                    Error in executing /var/tmp/stdin.127645
13  scripting  tst.gid.ksh                     Error in executing /var/tmp/stdin.127656
14  scripting  tst.projid.ksh                  Error in executing /var/tmp/stdin.127678
15  scripting  tst.sid.ksh                     Error in executing /var/tmp/stdin.127691
16  scripting  tst.taskid.ksh                  Error in executing /var/tmp/stdin.127708
17  scripting  tst.uid.ksh                     Error in executing /var/tmp/stdin.127721
18  ustack     tst.depth.ksh                   dtrace: failed to open output file 'out.138919': Permission denied
19  ustack     tst.spin.ksh                    dtrace: failed to open output file 'out.138923': Permission denied
```

### CODE REVIEW

**Comment by ahrens** [comment]
Reviewed at Delphix as part of a larger work which includes `while`, `entry->`, and `callers[]` syntactic sugar (some of which will be the subject of future pull requests), by @sebroy, @pcd1193182, and @ahl.

File: usr/src/lib/libdtrace/Makefile.com [hunk 1]
```diff
@@ -38,6 +38,7 @@ LIBSRCS = \
 	dt_dof.c \
 	dt_error.c \
 	dt_errtags.c \
+	dt_sugar.c \
```
Comment (ahl) [inline usr/src/lib/libdtrace/Makefile.com]: cute...

File: usr/src/lib/libdtrace/common/dt_parser.h [hunk 2]
```diff
@@ -27,6 +27,10 @@
  * Copyright (c) 2013 Joyent, Inc. All rights reserved.
  */
 
+/*
+ * Copyright (c) 2014 by Delphix. All rights reserved.
```
Comment (ahl) [inline usr/src/lib/libdtrace/common/dt_parser.h]: you've got another Delphix copyright above. Can you verify that it's just one per file?

**Comment by rmustacc** [comment]
@ahrens I spent some time going through it and it generally looks good.

File: usr/src/lib/libdtrace/common/dt_parser.h (hunk 2 above)
Comment (ahrens) [inline usr/src/lib/libdtrace/common/dt_parser.h]: I've fixed this, and looked for similar errors in the other modified files.  They all looked OK.

**Comment by ahrens** [comment]
Tested again with latest code, results still look good to me:
```
#   KIND       TEST                            DETAILS
0   cpc        tst.allcpus.ksh
1   dtraceUtil tst.ELFGenerationOut.d.ksh      dtrace: failed to link script /dev/stdin: failed to open d.out: Permission deni
2   dtraceUtil tst.ELFGenerationWithO.d.ksh    dtrace: failed to link script /dev/stdin: failed to open outputFile: Permission
3   funcs      tst.system.d                    sh: line 1: ping: not found
4   ip         tst.ipv4localicmp.ksh           dtrace: failed to execute /usr/sbin/ping: cannot find executable file
5   ip         tst.ipv4localudp.ksh            dtrace: failed to execute /usr/sbin/ping: cannot find executable file
6   ip         tst.ipv4remoteicmp.ksh          Couldn't run ping: No such file or directory
7   ip         tst.ipv4remotetcp.ksh           Couldn't run ping: No such file or directory
8   ip         tst.ipv4remoteudp.ksh           Couldn't run ping: No such file or directory
9   ip         tst.ipv6localicmp.ksh           dtrace: failed to execute /usr/sbin/ping: cannot find executable file
10  ip         tst.ipv6remoteicmp.ksh          remote IPv6 host.
11  ip         tst.localtcpstate.ksh
12  ip         tst.remotetcpstate.ksh          Couldn't run ping: No such file or directory
13  java_api   tst.Bean.ksh                    failed to create TestBean.out
```

**Comment by rmustacc** [comment]
Looks like a bunch of these tests would probably pass if the ping command was installed. Though some of the cpc failures are things I've seen before. I haven't seen the dtraceUtil failures, but I'd also be surprised if they were directly related to this.

**Comment by prakashsurya** [comment]
> Looks like a bunch of these tests would probably pass if the ping command was installed.

If needed, I can tweak the build code to install another package on the omnios system we use to build and test. Let me know if I should look into that.

**Comment by ahrens** [comment]
FYI, I installed `ping` and re-ran the tests.  The tcp/ip tests still fail for other reasons, but it fixed funcs/tst.system.d

```
#   KIND       TEST                            DETAILS
0   cpc        tst.allcpus.ksh
1   dtraceUtil tst.ELFGenerationOut.d.ksh      dtrace: failed to link script /dev/stdin: failed to open d.out: Permission den
2   dtraceUtil tst.ELFGenerationWithO.d.ksh    dtrace: failed to link script /dev/stdin: failed to open outputFile: Permissio
3   ip         tst.ipv4localicmp.ksh
4   ip         tst.ipv4localtcp.ksh
5   ip         tst.ipv4remoteicmp.ksh
6   ip         tst.ipv4remotetcp.ksh
7   ip         tst.ipv6remoteicmp.ksh          remote IPv6 host.
8   ip         tst.localtcpstate.ksh
9   ip         tst.remotetcpstate.ksh
10  java_api   tst.Bean.ksh                    failed to create TestBean.out
11  pid        tst.newprobes.ksh               dtrace: pid 452578 has exited
12  preprocessor err.ifdefnotendif.d
13  ustack     tst.depth.ksh                   dtrace: failed to open output file 'out.484059': Permission denied
14  ustack     tst.spin.ksh                    dtrace: failed to open output file 'out.484063': Permission denied```

**Comment by ahrens** [comment]
pushed to illumos

**Comment by zettamail** [comment]
Great stuff. Is there some documentation about this available somewhere?
I skimmed the change and didn't notice any manpage updates, so maybe
it's documented elsewhere?

Cheers,
-- 
Saso

**Comment by rmustacc** [comment]
There's not been an update to the DTrace guide. I'll see if we can put something together, though if you want to write some content there, I'd be happy to get that taken care of and updated.

**Comment by ahrens** [comment]
The dtrace features / language aren't documented in manpages.  Until we get the Dtrace guide updated, you can learn more about this feature from my talk at dtrace.conf(2016).  Slides: http://www.slideshare.net/MatthewAhrens/experimental-dtrace  and video: https://www.joyent.com/about/events/2016/dtrace-conf/videos/d-syntactic-sugar

**Comment by skiselkov** [comment]
Thanks.

**Pull request closed without merging**